networkx/algorithms/isomorphism/tree_isomorphism.py

```diff
@@ -1,29 +1,1021 @@
+
+    Given a list of lists whose entries are numbers between 0 and m. Define a
+    LENGTH map such that LENGTH(i) are the lists of S with length i.
+
+    For example, given ((1), (1, 2, 3), (1, 2, 2)), NONEMPTY(3) = [(1, 2, 3),
+    (1, 2, 2)].
+
+    Parameters
+    ----------
+    S : list of lists of ints (natural numbers)
+        The list of lists of natural numbers.
+
+    Returns
+    -------
+    LENGTH : dict of int: list of lists of ints
+        The LENGTH map such that LENGTH(i) are the lists of S with length i.
+
+    Notes
+    -----
+    This algorithm runs in O(|S|) time and space.
+
+    """
+    # Define an empty LENGTH map.
+    LENGTH = {}
+
+    # Traverse the lists in S and add them to the LENGTH.
+    for s in S:
+        if len(s) in LENGTH:
+            LENGTH[len(s)].append(s)
+        else:
+            LENGTH[len(s)] = [s]
+
+    return LENGTH
+
+
+def sort_lists_of_naturals(S):
+    """Lexicographically sort a list of lists of natural numbers.
+
+    Given a list of lists whose entries are numbers between 0 and m. Perform an
+    algorithm similar to bucket sort to return a sorted list with the elements
+    of S. The expected order is a lexicographical order.
+
+    Parameters
+    ----------
+    S : list of lists of ints (natural numbers)
+        The list of lists of natural numbers.
+
+    Returns
+    -------
+    sorted_list : list of lists of ints (natural numbers)
+        A lexicographically sorted list with the elements of S.
+
+    Notes
+    -----
+    This algorithm runs in O(l_total + m) time and O(l_total) space, where
+    l_total is the sum of the lengths of the lists in S and m is the max value
+    of all the lists in S.
+
+    """
+    # Get the maximum value contained in S.
+    m = max([max(s) for s in S])
+
+    # Define a list that will behave like a queue.
+    Q = []
+
+    # Define an empty array of size m whose entries are empty queues.
+    A = [[] for i in range(m + 1)]
+
+    # The max length found in S.
+    l_max = max([len(s) for s in S])
+
+    # Obtain the NONEMPTY and LENGTH maps.
+    NONEMPTY = categorize_entries(S)
+    LENGTH = categorize_lists(S)
+
+    l = l_max
+    while l >= 1:
+        # All of the elements in Q have length greater than l. For each element
+        # s in Q, append q to the queue A[s[l]].
+        while len(Q) > 0:
+            s = Q.pop()
+            A[s[l - 1]].append(s)
+
+        # Add the lists s of size l to the queue A[s[l]] if there are any.
+        if l in LENGTH:
+            for s in LENGTH[l]:
+                A[s[l - 1]].append(s)
+
+        # Sort the elements of A according to its l-th entry.
+        for i in NONEMPTY[l - 1]:
+            while len(A[i]) > 0:
+                s = A[i].pop()
+                Q.append(s)
+
+        # Proceed with the lists of size l-1.
+        l = l - 1
+
+    # Q now contains all of the elements of S sorted lexicographically.
+    return Q
+
+
+def get_levels(T, root, height):
```

Comment:
Each of these `get_` functions (there are around 5 of them) run its own BFS for the task; wouldn't it be faster to run one BFS and compute everything at once ?